core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -3686,4 +4157,71 @@ Sql assertRowsUnique(boolean[] ignoreNulls, Matcher<Boolean> matcher,
     }
 
   }
+
+  /**
+   * A configuration that describes how metadata should be configured.
+   */
+  public static class MetadataConfig {
+
+    private final String name;
+    private final Function<RelMetadataProvider, MetadataHandlerProvider> converter;
+    private final Supplier<RelMetadataProvider> defaultProviderSupplier;
+    private final boolean isCaching;
```

Comment:
Caching is required volcano planner.

Reply:
Can you point me to where that is? That seems like it should be a bug in volcano, not a requirement.

Reply:
I'm wondering if you're referring of the requirement to ensure we're breaking cycles (as opposed to actually caching). Do you know if there are any tests to verify this behavior (cycle-break)?

Reply:
Yes, I am refering to cycle detection.  Sorry, about not being clearer.  I am not sure if their are any tests in calcite that trigger it.

Reply:
I've just added a test to verify no cycles in both impls.